### PR TITLE
adapter: collect storage usage by shard

### PR DESCRIPTION
**Pull request by benesch**
Move the `mz_catalog.mz_storage_usage` table to
`mz_internal.mz_storage_usage_by_shard`, and include storage usage by
shard ID.

A future commit will add back `mz_catalog.mz_storage_usage` as a view
that joins `mz_storage_usage_by_shard` with the forthcoming table that
maps shard IDs and object IDs together (https://github.com/MaterializeInc/materialize/pull/15052).

Touches https://github.com/MaterializeInc/cloud/issues/3726.

### Motivation

  * This PR makes progress on a known-desirable feature.

### Tips for reviewer

@teskje—can you review the first commit, which changes the logic for determining whether a relation is unstable?

@sploiselle or @jpepin—can you review the second commit?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: test/testdrive/catalog.td
Comment (teskje) [inline test/testdrive/catalog.td:487]: These need to be added to the `> SHOW TABLES FROM mz_internal` test below.

File: doc/user/content/sql/system-catalog.md
Comment (teskje) [inline doc/user/content/sql/system-catalog.md:328]: Wait, this is `mz_views`, not `mz_view_keys`!

**Review by teskje — CHANGES_REQUESTED** [review]
The code changes in commit 1 look good! The tests and user docs still need tweaking.

**Review by teskje — APPROVED** [review]
LGTM!

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (jpepin) [inline src/adapter/src/catalog/builtin_table_updates.rs:792]: Nit: Could rename the `object_id` vars here to `shard_id` as well, but nbd

**Review by jpepin — APPROVED** [review]

File: src/adapter/src/catalog/builtin_table_updates.rs
Comment (benesch) [reply src/adapter/src/catalog/builtin_table_updates.rs]: Good callout, thanks! I tightened up this whole function.

**Comment by sploiselle** [comment]
@benesch fixed merge conflicts and setting to auto merge so I can work on top of this in #15052

**Comment by benesch** [comment]
Ahh sorry @sploiselle but I'm 98%+ sure that this is a backincompat change and needs to wait until next Tue afternoon to merge.

**Comment by benesch** [comment]
Can you rebase your PR on top of this one, and then I'll close this one and we'll merge it all together in yours?

**Comment by sploiselle** [comment]
@benesch Whoops! Glad you're always online :D I can rebase on top of this and place it in my branch, no sweat

**Comment by benesch** [comment]
Folded into https://github.com/MaterializeInc/materialize/pull/15052.

**Pull request closed without merging**
